packager/src/Resolver/polyfills/require.js

```diff
@@ -155,23 +155,33 @@ function loadModuleImplementation(moduleId, module) {
   // factory to keep any require cycles inside the factory from causing an
   // infinite require loop.
   module.isInitialized = true;
-  const exports = module.exports = {};
+  module.exports = {};
   const {factory, dependencyMap} = module;
   try {
     if (__DEV__) {
       // $FlowFixMe: we know that __DEV__ is const and `Systrace` exists
       Systrace.beginEvent('JS_require_' + (module.verboseName || moduleId));
     }
 
-    const moduleObject: Module = {exports};
+    const moduleObject = {
+      get id() {
+        return moduleId;
+      },
+      get exports() {
+        return module.exports;
+      },
+      set exports(value) {
+        module.exports = value;
+      }
+    };
     if (__DEV__ && module.hot) {
       moduleObject.hot = module.hot;
     }
 
     // keep args in sync with with defineModuleCode in
     // packager/src//Resolver/index.js
     // and packager/src//ModuleGraph/worker.js
-    factory(global, require, moduleObject, exports, dependencyMap);
+    factory.call(module.exports, global, require, moduleObject, module.exports, dependencyMap);
```

Comment:
I am aware that node.js does this. Are you aware of any modules that actually depend on this? We will have to run some internal perf tests on this to gauge real-world impact.